go/pkg/cas/ioutil.go

```diff
@@ -1,39 +1,85 @@
 package cas
 
 import (
+	"bufio"
 	"bytes"
 	"fmt"
 	"io"
+	"os"
+	"sync"
 )
 
 // This file contains various IO utility types and functions.
 
-// readSeekCloser can Read, Seek and Close.
-// TODO(nodir): use io.ReadSeekCloser after dropping Go 1.15 support.
-type readSeekCloser interface {
-	io.ReadSeeker
+// byteSource is a common interface to in-memory blobs and files.
+// SeekStart is like io.Seeker.Seek, but only supports io.SeekStart.
+type byteSource interface {
+	io.Reader
 	io.Closer
+	SeekStart(offset int64) error
```

Comment:
We might need to preserve offsets for retries?

Reply:
Sorry, could you elaborate? Isn't preserving offsets the job of the caller?

Reply:
My bad again - I read `SeakStart()` without an offset, oops.
Nvm my comment.